cmd/access/node_builder/staked_access_node_builder.go

```diff
@@ -24,14 +27,44 @@ func NewStakedAccessNodeBuilder(anb *FlowAccessNodeBuilder) *StakedAccessNodeBui
 	}
 }
 
+func (fnb *StakedAccessNodeBuilder) InitIDProviders() {
+	fnb.Module("id providers", func(builder cmd.NodeBuilder, node *cmd.NodeConfig) error {
+		idCache, err := p2p.NewProtocolStateIDCache(node.Logger, node.State, fnb.ProtocolEvents)
+		if err != nil {
+			return err
+		}
+
+		fnb.IdentityProvider = idCache
+		// translator
+		fnb.SyncEngineParticipantsProvider = id.NewFilteredIdentifierProvider(
+			filter.And(
+				filter.HasRole(flow.RoleConsensus),
+				filter.Not(filter.HasNodeID(node.Me.NodeID())),
+				p2p.NetworkingSetFilter,
+			),
+			idCache,
+		)
+		fnb.IDTranslator = p2p.NewHierarchicalIDTranslator(idCache, p2p.NewUnstakedNetworkIDTranslator())
+
+		// TODO: NetworkingIdentifierProvider should be the same as the one used in scaffold.go if this AN
+		// doesn't participate in unstaked network.
+		// If it does, then we can just use the default one (peerstoreProvider)
```

Comment:
@vishalchangrani when we remove the second network later.

Reply:
but the peerstoreProvider assumes that peers will first initiate and connect to this node. For the staked access node which also participates in the unstaked network, the staked consensus nodes and staked collection nodes may or may not connect. There happens a random selection of other nodes to which a node connects to at startup.

We would need a combination of the `peerStoreProvider` and the `ProtocolStateIDCache` for a staked node which also participates in the unstaked network

Reply:
So elsewhere in my PR, i ensure that the protocol state nodes are *ALWAYS* present in the peerstore.

You can search for a method called `UpdateNodeAddresses` or something like that.

In other words, we can rely on the peerstore because we explicitly put the addresses of all staked nodes in the peerstore, and set those addresses to be permanent until the next epoch.

Reply:
I see it. thanks.

Reply:
[Yup](https://github.com/onflow/flow-go/pull/1133/files#diff-95120943dbf1b8334da6c35dc229d380cea1f66bb6751a652aab627b9082ffa8R203-R227)